// lisk-framework/dist-node/testing/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mocks = exports.fixtures = exports.getAccountSchemaFromModules = exports.getModuleInstance = exports.createGenesisBlock = exports.createTransaction = void 0;
const fixtures = require("./fixtures");
exports.fixtures = fixtures;
const mocks = require("./mocks");
exports.mocks = mocks;
var create_transaction_1 = require("./create_transaction");
Object.defineProperty(exports, "createTransaction", { enumerable: true, get: function () { return create_transaction_1.createTransaction; } });
var create_genesis_block_1 = require("./create_genesis_block");
Object.defineProperty(exports, "createGenesisBlock", { enumerable: true, get: function () { return create_genesis_block_1.createGenesisBlock; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getModuleInstance", { enumerable: true, get: function () { return utils_1.getModuleInstance; } });
Object.defineProperty(exports, "getAccountSchemaFromModules", { enumerable: true, get: function () { return utils_1.getAccountSchemaFromModules; } });
__exportStar(require("./create_block"), exports);
__exportStar(require("./app_env"), exports);
__exportStar(require("./create_contexts"), exports);
__exportStar(require("./block_processing_env"), exports);
